File: 144.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ret;
        
        TreeNode *cur = root;
        vector<TreeNode *>st;
        
        while (!st.empty() || cur != nullptr) {
            while (cur != nullptr) {
                ret.push_back(cur->val);
                st.push_back(cur);
                cur = cur->left;
            }
            
            cur = st.back();
            st.pop_back();
            
            cur = cur->right;
        }
        
        return ret;
    }
};

File: go/144.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func preorderTraversal(root *TreeNode) []int {
	var ret = make([]int, 0)
	if root == nil {
		return ret
	}
	var stack = make([]*TreeNode, 0)
	var cur *TreeNode = root

	for len(stack) > 0 || cur != nil {
		for cur != nil {
			ret = append(ret, cur.Val)
			stack = append(stack, cur)
			cur = cur.Left
		}

		cur = stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if cur != nil {
			cur = cur.Right
		}
	}

	return ret
}
